### ISSUE TITLE
Decoding and Encoding a Non-Empty Array

### ISSUE THREAD

**Issue by cjduncana** [basti1302/elm-non-empty-array#4]
I propose to add two new functions to `NonEmptyArray` to decode and encode a non-empty array.

```elm
decoder : Decoder a -> Decoder (NonEmptyArray a)
decoder =
    Decode.list
        >> Decode.andThen
            (\list ->
                case fromList list of
                    Just nea ->
                        Decode.succeed nea

                    Nothing ->
                        Decode.fail "Decoded an empty list"
            )

encode : NonEmptyArray Value -> Value
encode =
    toList
        >> Encode.list
```

The decoder function can be simplified even more using `Maybe` helper functions like `map` and `withDefault`, but you can see the idea with the code provided.

**Comment by basti1302** [comment]
Sorry for the late reply. I'm generally in favor of adding this functionality to this package. The function `decoder` looks exactly right (it also looks suspiciously similar to the code I have written repeatedly in a number of projects that use non-empty-array ;-) ).

It might be nice to have a second variant  `decoderWithDefault : Decoder a -> a -> Decoder (NonEmptyArray a)` that takes a default value and, instead of failing on an empty list, returns an array with that default element. (This is nice for the use case where you want to automatically create a new empty-ish element for the user in case the backend did not provide one.)

I'm not sure about the `encode` function. I think it would make more sense to let the user specify an encoding function for a single element and handle the mapping over the NEA ourselves:

```
encode : (a -> Value) -> NonEmptyArray a -> Value
encode encodeSingle nea =
    map encodeSingle nea
    |> toList
    |> Encode.list
```

What do you think?

Also, would you be willing to create a PR for this?

**Comment by basti1302** [comment]
This has landed in version 2.1.0.
